### PR TITLE
Added support for the MPSSE, SPI and GPIO

### PR DESCRIPTION
**Pull request by JulianS-Uni**
The modern modules FT232H, FT2232H and FT4232H do have one or more MPSSE units which can be programmed to create different bus systems. I added a class (MPSSE) which implements all the functions of the MPSSE. On top of this i added a class (SPI) which sets up the MPSSE to work as a SPI master. From a SPI master it is possible to create spiSlaves which are for the actual communication with other devices. I also added a GPIO and an InputOutput class which allow to use the MPSSE as a simple digital pin extension.

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,5 @@
+{
+    "python.linting.pylintEnabled": true,
+    "python.linting.enabled": true,
+    "python.linting.mypyEnabled": false
+}
```
Comment (lsgunth) [inline .vscode/settings.json]: Please don't commit developer environment stuff. Or at the very least, put it in a separate PR for it's own consideration.

File: ft232/mpsse.py [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+    All commands are described in this reference sheet 
+    https://www.ftdichip.com/Support/Documents/AppNotes/AN_108_Command_Processor_for_MPSSE_and_MCU_Host_Bus_Emulation_Modes.pdf
+    """
+    def __init__(self, Ft2xx=Ft232("Ft232H")):
+        self.Ft232 = Ft2xx
+        self.PinDirections = 0b0000000000001011
+        self.PinValues = 0x0000
+
+    def _dataCommand(self, ReadWriteData, AsBitsBytes, WriteClkEdge, BitOrder, TmsOnOff):
+        """ Before writing or reading data a Commandbyte has to be sent in advance which defines how the data is sent or received and how it should be interpreted. After sending the Commandbyte the Databytes are expected.
```
Comment (lsgunth) [inline ft232/mpsse.py:42]: Please wrap all lines to 80 characters. It makes it much easier to read.

File: ft232/__init__.py [hunk 3]
```diff
@@ -18,13 +18,11 @@
 try:
     from .d2xx import list_devices, D2XXException
     from .d2xx import D2xx as Ft232
-
-    Ft232Exception = d2xx.D2XXException
+    from .d2xx import D2XXException as Ft232Exception
```
Comment (lsgunth) [inline ft232/__init__.py:21]: This is a nice clean up, but can you please submit it as a separate PR? It's unrelated to this PR.

File: ft232/d2xx.py [hunk 4]
```diff
@@ -283,7 +283,7 @@ def setWriteTimeout(self, timeout):
 
     def getWriteTimeout(self):
         """Get the current timeout setting."""
-        return self._writeTimeout
+        return self._writetimeout
```
Comment (lsgunth) [inline ft232/d2xx.py:279]: This looks like a bug that exists in d2xx and it looks like there's some more inconsistent usage of writetimeout. 

I'd appreciate it if you can fix all of it and send this as a separate PR.

File: ft232/d2xx.py [hunk 5]
```diff
@@ -362,7 +362,9 @@ def __repr__(self):
 
 
     def write(self, s):
-        buf = c.create_string_buffer(s)
+        ## Debug ##
+        print(s.hex())
```
Comment (lsgunth) [inline ft232/d2xx.py]: Please don't commit debug lines like this. Even if they are commented out.

File: ft232/libftdi.py [hunk 6]
```diff
@@ -206,7 +206,7 @@ def _setDataCharacteristics(self):
     def setByteSize(self, bytesize):
         """Change byte size."""
         if bytesize not in self.BYTESIZES: raise ValueError("Not a valid byte size: %r" % bytesize)
-        self._bytesize = bytesize0
+        self._bytesize = bytesize
```
Comment (lsgunth) [inline ft232/libftdi.py:202]: Ditto about the bug fix, please send it as a separate PR (it could be included as a second patch with the fix to the writetimeout).

**Review by lsgunth — CHANGES_REQUESTED** [review]
Thanks for the contribution!

I'd be more than happy to merge this but I'd really appreciate it if you can clean up the series for me a bit first to make it a little easier for me to review:

1) Please separate out any bug fixes for existing code into a seperate PR. I can review and merge them quicker and it removes from the noise I have to sort out in this PR.

2) Please squash all the patches that make changes to patches earlier in the series. It's very annoying to review code in one patch only to find that it's been changed, removed or fixed in a subsequent patch. 

3) I'd appreciate a little more description in the commit messages. "Some deletes" doesn't tell me much.

In case you don't know, git has a lot of tools for editing your commit history: see [rewriting history](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History). So all you need to do is use those tools to do some cleanup and then force push to your branch and it will automatically update this PR. Post a comment when you'd like some more review.

I've gone through some of the patches and made some nits below.

Logan

**Comment by JulianS-Uni** [comment]
Hi, i am happy you appreciate my work. I try to seperate my work into some smaller PR. But i dont really know much about git so ill try to my best.
